### DISCUSSION TITLE
Scene2D builders inside constructors

### DISCUSSION THREAD

**Discussion by Divelix** [libktx/ktx#306] [Q&A] [answered]
I have UI with several `Table`s that I want to put into separate classes, so I have code like that:
```kotlin
class MyTable: Table() {
    init {
        val subTable1 = scene2d.table { ... }
        val subTable2 = scene2d.table { ... }

        add(subTable1)
        row()
        add(subTable2)
    }
}
```
Can I use builders there somehow to get something like that?
```kotlin
class MyTable: Table() {
    init {
        table { ... } // subTable1
        row()
        table { ... } // subTable2
    }
}
```

**Comment by czyzby** [comment, marked as answer]
Yes, you have to implement the `KTable` interface:

```kotlin
import ktx.scene2d.KTable
import com.badlogic.gdx.scenes.scene2d.ui.Table

class MyTable: Table(), KTable {
  init {
    table { } // subTable1
    row()
    table { } // subTable2
  }
}
```

Since `MyTable` already extends `Table`, there aren't any extra methods that you have to override - all you have to do is to add the interface.

Similar interfaces exist for groups, trees, etc.

**Comment by Divelix** [comment]
Awesome! I had gut feelling that it is possible)

**Comment by czyzby** [comment]
I added a similar example to the README. Thanks.

**Comment by Divelix** [comment]
I just found out that in this case I can't use class fields, like:

![image](https://user-images.githubusercontent.com/25386431/92448544-6c872600-f1c1-11ea-87be-7fed8083cfec.png)

I faced this problem in #301, but in that case I just refused using `KGroup`. How can I fix that and keep using `KGroup`?

**Comment by czyzby** [comment]
@Divelix This is because `this` is changed via the DSL to prevent from accessing nested properties. If you need to access a class property within the constructor, add an explicit receiver: `this@ClassName.` - for example:

```kotlin
import com.badlogic.gdx.scenes.scene2d.Group

class Try: Group(), KGroup {
  val str = "text"
  
  init {
    table { 
      label(this@Try.str)
    }
  }
}
```

**Comment by czyzby** [comment]
Just wanted to say that this is an expected side effect of the Kotlin DSL builders, and it actually has some advantages. For example, you cannot immediately access properties of parent actors within the children blocks:

```kotlin
table {
  label("Child") {
    // This will not compile, so you will not modify the table accidentally:
    pad(2f)
    // If you really need to access parent's properties, you can do so explicitly:
    this@table.pad(2f)
  }

  // This compiles normally, as we're within the table block:
  pad(2f)
}
```
